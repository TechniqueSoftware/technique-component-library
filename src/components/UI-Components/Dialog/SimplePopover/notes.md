# SimplePopover
This component is an implementation of Material-UI's [Simple Popover demo](https://material-ui.com/components/popover/#simple-popover). 
The content of the popover as well as the target component used to open the popover is configurable. The component currently
makes an assumption that all popovers will have a Close button at the bottom and implements this by default. This will 
probably change in the future but can be refactored easily to handle customization of the DialogActions.

## JIRA
Related jira ticket: https://clubos.atlassian.net/browse/CR-674
