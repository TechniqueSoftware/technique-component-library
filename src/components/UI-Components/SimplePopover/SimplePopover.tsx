import * as React from 'react';
import Popover, { PopoverProps } from '@material-ui/core/Popover';

export interface SimplePopoverProps {
  targetComponent: React.ReactNode;
  PopoverProps?: Partial<PopoverProps>;
  children?: React.ReactNode;
}

function SimplePopover(props: SimplePopoverProps) {

  const [anchorEl, setAnchorEl] = React.useState(null);

  function handleClick(event: React.SyntheticEvent) {
    setAnchorEl(event.currentTarget);
  }

  function handleClose() {
    setAnchorEl(null);
  }

  const open = Boolean(anchorEl);

  const popoverProps: PopoverProps = {
    open,
    anchorEl,
    onClose: handleClose,
    anchorOrigin: {
      vertical: anchorEl ? anchorEl.getBoundingClientRect().height + 4 : 'bottom',
      horizontal: 'left',
    },
    transformOrigin: {
      vertical: 'top',
      horizontal: 'left',
    },
    ...props.PopoverProps,
  };

  return (
    <div>
      <div onClick={handleClick} data-test={'popover-target-wrapper'}>
        {props.targetComponent}
      </div>
      <Popover {...popoverProps}>
        {props.children}
      </Popover>
    </div>
  );
}

export default SimplePopover;
