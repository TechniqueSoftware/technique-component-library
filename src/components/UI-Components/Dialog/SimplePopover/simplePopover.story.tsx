import * as React from 'react';
import { createCUICStoriesOf } from '../../utils/utils';
import { MODULE_NAME } from '../../constants';
import clubOsTheme from '../../storyThemes';
import SimplePopover,{ SimplePopoverProps} from '../../SimplePopover/SimplePopover';
import ButtonWithIcon from '../../Buttons/ButtonWithIcon/ButtonWithIcon';
import { MuiThemeProvider } from '@material-ui/core/styles';
import DialogContent from '@material-ui/core/DialogContent';
import Box from '@material-ui/core/Box';

function render() {

  const simplePopoverProps: SimplePopoverProps = {
    targetComponent: <ButtonWithIcon>Target Component</ButtonWithIcon>
  };

  return (
    <div>
      <MuiThemeProvider theme={clubOsTheme}>
        <Box p={2}>
          <SimplePopover {...simplePopoverProps}>
            <DialogContent>
              <p>Any content can be added here</p>
            </DialogContent>
          </SimplePopover>
        </Box>
      </MuiThemeProvider>
    </div>);
}

createCUICStoriesOf(MODULE_NAME.DIALOGS, module)
  .addCUICStory({
    render,
    name: 'SimplePopover',
    notes: require('./notes.md').default
  });
