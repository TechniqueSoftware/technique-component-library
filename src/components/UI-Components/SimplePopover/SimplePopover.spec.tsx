import SimplePopover, { SimplePopoverProps } from './SimplePopover';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('SimplePopover', () => {

  function renderShallow(props?: Partial<SimplePopoverProps>) {
    const baseProps: SimplePopoverProps = {
      targetComponent: <div>Mock target component</div>,
      ...props
    };

    return shallow(<SimplePopover {...baseProps} />);
  }

  test('should render with base props', () => {
    const shallowWrapper = renderShallow();
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

  test('should set popover to open when target wrapper is clicked', () => {
    const shallowWrapper = renderShallow();
    const target = shallowWrapper.find('[data-test="popover-target-wrapper"]').first();
    const currentTarget = {
      getBoundingClientRect: jest.fn(() => ({
        height: 999
      }))
    };
    target.simulate('click', { currentTarget });
    const popoverWrapper: any = shallowWrapper.find('WithStyles(ForwardRef(Popover))');
    expect(popoverWrapper.props().open).toBe(true);
    expect(popoverWrapper.props().anchorEl).toBe(currentTarget);
  });
});
