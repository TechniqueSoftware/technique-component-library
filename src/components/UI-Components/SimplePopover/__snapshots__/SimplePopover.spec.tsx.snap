// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimplePopover should render with base props 1`] = `
"<div>
  <div onClick={[Function: handleClick]} data-test=\\"popover-target-wrapper\\">
    <div>
      Mock target component
    </div>
  </div>
  <WithStyles(ForwardRef(Popover)) open={false} anchorEl={{...}} onClose={[Function: handleClose]} anchorOrigin={{...}} transformOrigin={{...}} />
</div>"
`;
